### PR TITLE
Add the `ios run` command

### LINKED ISSUE #365: Support running iOS apps without a TestRunner
**Context**
Some iOS apps we will deploy won't necessarily have the TestRunner inside. For these cases, XHarness currently exits with a non-zero exit code as it never receives test results:
```
info: Application finished the test run successfully with some failed tests
info: No test log file was produced
XHarness exit code: 1
```

Helix SDK also exits when it doesn't detect any XML results:
https://github.com/dotnet/arcade/blob/a9a80fb35d2e7da21509441d665a40022ce8f1b4/src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/ios-helix-job-payload.sh#L130-L139

**Goal**
- Add a new command **`xharness ios run`** that will not expect TestRunner in the running app and will rather just run it and return the same exit code as was returned by the iOS app. We need #364 for this
- Update docs to mention this
- Reuse as much of the existing `ios test` command and of the `AppRunner` as possible

### PR DESCRIPTION
**Pull request by premun**
Adds a new command that runs iOS app without a TestRunner and tries to detect its exit code.
- `AppRunner` renamed to `AppTester`
- A new `AppRunner` added for the new command
- Both these classes and their command/arguments share a fair amount of code now

Integration tests will follow later to unblock other work at the moment.

Resolves #365 #363

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -99,7 +105,9 @@ These files are:
 
 ## Test Runners
 
-The repository also contains several TestRunners that are bundled inside of the application and execute the tests.
+The repository also contains several TestRunners which are libraries that can be bundled inside of the application and execute the tests.
+The TestRunner detects and exeuctes unit tests inside of the application. It also connects to XHarness over TCP connection from within the running app bundle and reports test run results/state.
```
Comment (MaximLipnin) [inline README.md:109]:
```suggestion
The TestRunner detects and executes unit tests inside of the application. It also connects to XHarness over TCP connection from within the running app bundle and reports test run results/state.
```

File: src/Microsoft.DotNet.XHarness.CLI/CommandArguments/AppRunCommandArguments.cs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            {
+                "app|a=", "Path to already-packaged app",
+                v => AppPackagePath = RootPath(v)
+            },
+            {
+                "output-directory=|o=", "Directory in which the resulting package will be outputted",
+                v => OutputDirectory = RootPath(v)
+            },
+            {
+                "targets=|t=", "Comma-delineated list of targets to test for",
```
Comment (MaximLipnin) [inline src/Microsoft.DotNet.XHarness.CLI/CommandArguments/AppRunCommandArguments.cs:57]:
```suggestion
                "targets=|t=", "Comma-delimited list of targets to test for",
```

File: src/Microsoft.DotNet.XHarness.CLI/CommandArguments/iOS/iOSRunCommandArguments.cs [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+using System.IO;
+using System.Linq;
+using Microsoft.DotNet.XHarness.iOS.Shared;
+using Microsoft.DotNet.XHarness.iOS.Shared.Utilities;
+using Mono.Options;
+
+namespace Microsoft.DotNet.XHarness.CLI.CommandArguments.iOS
+{
+    /// <summary>
+    /// Specifies the channel that is used to comminicate with the device.
```
Comment (MaximLipnin) [inline src/Microsoft.DotNet.XHarness.CLI/CommandArguments/iOS/iOSRunCommandArguments.cs:16]:
```suggestion
    /// Specifies the channel that is used to communicate with the device.
```

File: src/Microsoft.DotNet.XHarness.CLI/Commands/Android/AndroidTestCommand.cs [hunk 4]
```diff
@@ -49,7 +50,7 @@ internal class AndroidTestCommand : TestCommand
 Arguments:
 ";
 
-        public AndroidTestCommand() : base(CommandHelp)
+        public AndroidTestCommand() : base("test", true, CommandHelp)
```
Comment (greenEkatherine) [inline src/Microsoft.DotNet.XHarness.CLI/Commands/Android/AndroidTestCommand.cs:53]: The second argument was false before, could you please explain, why did you change it?

File: src/Microsoft.DotNet.XHarness.iOS.Shared/Hardware/SimulatorLoader.cs [hunk 5]
```diff
@@ -462,6 +462,32 @@ public ISimulatorDevice FindCompanionDevice(ILog log, ISimulatorDevice device)
             return _availableDevices.Single(v => v.UDID == pair.Companion);
         }
 
+        public async Task<(ISimulatorDevice Simulator, ISimulatorDevice? CompanionSimulator)> FindSimulators(TestTargetOs target, ILog log, int retryCount, bool createIfNeeded = true, bool minVersion = false)
```
Comment (greenEkatherine) [inline src/Microsoft.DotNet.XHarness.iOS.Shared/Hardware/SimulatorLoader.cs:465]: It seems that you don't use `retryCount` parameter. Do you need it?

File: src/Microsoft.DotNet.XHarness.CLI/Commands/Android/AndroidTestCommand.cs (hunk 4 above)
Comment (premun) [reply src/Microsoft.DotNet.XHarness.CLI/Commands/Android/AndroidTestCommand.cs:53]: Ah, good catch!

File: src/Microsoft.DotNet.XHarness.iOS.Shared/Hardware/SimulatorLoader.cs (hunk 5 above)
Comment (premun) [reply src/Microsoft.DotNet.XHarness.iOS.Shared/Hardware/SimulatorLoader.cs:465]: Also a good catch :)

File: src/Microsoft.DotNet.XHarness.CLI/Commands/iOS/iOSTestCommand.cs [hunk 6]
```diff
@@ -28,337 +22,129 @@ namespace Microsoft.DotNet.XHarness.CLI.Commands.iOS
...
-            }
-
-            return exitCode;
-        }
+        protected override iOSRunCommandArguments iOSRunArguments => _arguments;
 
-        private static bool IsLldbEnabled() => File.Exists(s_mlaunchLldbConfigFile);
-
-        private void NotifyUserLldbCommand(ILogger logger, string line)
+        public iOSTestCommand() : base("test", true, CommandHelp)
```
Comment (greenEkatherine) [inline src/Microsoft.DotNet.XHarness.CLI/Commands/iOS/iOSTestCommand.cs:35]:
```suggestion
        public iOSTestCommand() : base("test", false, CommandHelp)
```

File: src/Microsoft.DotNet.XHarness.CLI/Commands/iOS/iOSRunCommand.cs [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    {
+        private const string CommandHelp = "Runs a given iOS/tvOS/watchOS application bundle in a target device/simulator and tries to detect exit code (might not work reliably across iOS versions).";
+
+        private readonly iOSRunCommandArguments _arguments = new iOSRunCommandArguments();
+
+        protected override iOSRunCommandArguments iOSRunArguments => _arguments;
+        protected override string CommandUsage { get; } = "ios run [OPTIONS] [-- [RUNTIME ARGUMENTS]]";
+        protected override string CommandDescription { get; } = CommandHelp;
+
+        public iOSRunCommand() : base("run", true, CommandHelp)
```
Comment (greenEkatherine) [inline src/Microsoft.DotNet.XHarness.CLI/Commands/iOS/iOSRunCommand.cs:32]:
```suggestion
        public iOSRunCommand() : base("run", false, CommandHelp)
```

File: src/Microsoft.DotNet.XHarness.CLI/Commands/Android/AndroidTestCommand.cs (hunk 4 above)
Comment (greenEkatherine) [reply src/Microsoft.DotNet.XHarness.CLI/Commands/Android/AndroidTestCommand.cs]: Thanks, I suggested two more changes based on that

File: src/Microsoft.DotNet.XHarness.iOS.Shared/Hardware/SimulatorLoader.cs [hunk 8]
```diff
@@ -462,6 +462,31 @@ public ISimulatorDevice FindCompanionDevice(ILog log, ISimulatorDevice device)
...
+            {
+                try
+                {
+                    return await FindSimulators(target, log);
+                }
+                catch (Exception e)
+                {
+                    log.WriteLine($"Failed to find/create simulator (attempt {attempt}/{retryCount}):" + Environment.NewLine + e);
+
+                    if (attempt == retryCount)
```
Comment (greenEkatherine) [inline src/Microsoft.DotNet.XHarness.iOS.Shared/Hardware/SimulatorLoader.cs:478]: Could `retryCount` be 0?

File: src/Microsoft.DotNet.XHarness.CLI/Commands/iOS/iOSRunCommand.cs (hunk 7 above)
Comment (premun) [reply src/Microsoft.DotNet.XHarness.CLI/Commands/iOS/iOSRunCommand.cs:32]: This was actually `true` before and needs to be `true`

File: src/Microsoft.DotNet.XHarness.CLI/Commands/iOS/iOSTestCommand.cs (hunk 6 above)
Comment (premun) [reply src/Microsoft.DotNet.XHarness.CLI/Commands/iOS/iOSTestCommand.cs:35]: This was actually `true` before and needs to be `true`

File: src/Microsoft.DotNet.XHarness.iOS.Shared/Hardware/SimulatorLoader.cs (hunk 8 above)
Comment (premun) [reply src/Microsoft.DotNet.XHarness.iOS.Shared/Hardware/SimulatorLoader.cs:478]: Good catch

Comment (greenEkatherine) [reply src/Microsoft.DotNet.XHarness.iOS.Shared/Hardware/SimulatorLoader.cs:478]: Sorry, did you add the change to handle that case? I cannot find it

Comment (premun) [reply src/Microsoft.DotNet.XHarness.iOS.Shared/Hardware/SimulatorLoader.cs:478]: Oops, committed to a wrong branch! Thanks for noticing

File: src/Microsoft.DotNet.XHarness.CLI/Commands/iOS/iOSTestCommand.cs (hunk 6 above)
Comment (greenEkatherine) [reply src/Microsoft.DotNet.XHarness.CLI/Commands/iOS/iOSTestCommand.cs:35]: I double check it and still it seems to me that it was false by default. Please correct me if I'm wrong:
image

Comment (premun) [reply src/Microsoft.DotNet.XHarness.CLI/Commands/iOS/iOSTestCommand.cs]: You're right! I am sorry, it was the WASM command that has it set to true. I didn't check it properly and thought the base class was setting this to `true`. Fixed

**Review by greenEkatherine — APPROVED** [review]

**Review by akoeplinger — APPROVED** [review]
